pig/src/org/partiql/pig/generator/kotlin/KTypeDomainConverter.kt

```diff
@@ -32,29 +33,34 @@ import org.partiql.pig.domain.model.TypeUniverse
 import org.partiql.pig.util.snakeToCamelCase
 import org.partiql.pig.util.snakeToPascalCase
 
-internal fun TypeUniverse.convertToKTypeUniverse(): KTypeUniverse {
+internal fun TypeUniverse.convertToKTypeUniverse(domains: Set<String>?): KTypeUniverse {
     val allTypeDomains: List<TypeDomain> = this.computeTypeDomains()
 
-    val kotlinTypeDomains: List<KTypeDomain> = allTypeDomains.map { typeDomain ->
-        KTypeDomainConverter(typeDomain).convert()
-    }
+    val kotlinTypeDomains: List<KTypeDomain> = allTypeDomains.filterDomains(domains)
+        .map { typeDomain -> KTypeDomainConverter(typeDomain).convert() }
 
-    val allKTransforms = this.statements.filterIsInstance<Transform>().map { dt ->
-        // TODO: handle properly if one of the calls to .single fails below...
-        val sourceDomain = allTypeDomains.single { it.tag == dt.sourceDomainTag }
-        val destDomain = allTypeDomains.single { it.tag == dt.destinationDomainTag }
-        val difference = sourceDomain.computeTransform(destDomain)
+    val allKTransforms = this.statements.filterIsInstance<Transform>()
 
-        val converter = KTypeDomainConverter(difference)
-        val sourceDomainDifference = converter.convert()
+    val sourceDomains = allKTransforms.map { dt -> allTypeDomains.singleOrNull { it.tag == dt.sourceDomainTag } }
+    val destDomains = allKTransforms.map { dt -> allTypeDomains.singleOrNull { it.tag == dt.destinationDomainTag } }
 
-        KTransform(
-            sourceDomainDifference = sourceDomainDifference,
-            destDomainKotlinName = destDomain.tag.snakeToPascalCase()
-        )
-    }
+    val validKTransforms = sourceDomains.zip(destDomains)
+        .filter { (sourceDomain, destDomain) -> sourceDomain != null && destDomain != null }
+        // sourceDomain and destDomain should both be non-null here, we just checked
+        .filter { (_, destDomain) -> domains?.let { destDomain!!.tag in it } ?: true }
```

Comment:
if you use `.mapNotNull` on lines 44 and 45 instead of `.map` you can drop the filter on line 45 and the `!!`s below.

Reply:
Even worse, you can end up in a situation where the `zip`ped lists won't have the same length:

```
sourceDomains [null, null, srcDomain3]
destDomains = [null, dstDomain2, dstDomain3]
```